development_bank_wales/pipeline/feature_preparation/feature_engineering.py

```
@@ -0,0 +1,327 @@
+    """Get difference in energy efficiency between first to latest entry for all categories.
+
+    Args:
+        df1 (pd.DataFrame): Earliest/first property records, including ENERGY_EFF features.
+        df2 (pd.DataFrame): Latest property records, including ENERGY_EFF features.
+        keep (str, optional): Which to keep: earliest/first or latest. Defaults to "first".
+        identifier (str, optional): Unique property identifier. Defaults to "UPRN".
+
+    Returns:
+        df (pd.DataFrame): Dataframe with difference in energy efficiency for all categories.
+    """
+
+    quality_dict = {"Very Good": 5, "Good": 4, "Average": 3, "Poor": 2, "Very Poor": 1}
+
+    for cat in cats:
+
+        # Transform efficiency label to numeric score and compute difference
+        eff_feature = "{}_ENERGY_EFF".format(cat)
+        df1[eff_feature + "_NUM"] = df1[eff_feature].map(quality_dict)
+        df2[eff_feature + "_NUM"] = df2[eff_feature].map(quality_dict)
+
+        combo = pd.merge(
+            df1[[eff_feature + "_NUM", identifier]],
+            df2[[eff_feature + "_NUM", identifier]],
+            on=identifier,
+        )
+
+        combo[cat + "_EFF_DIFF"] = (
+            combo[eff_feature + "_NUM_y"] - combo[eff_feature + "_NUM_x"]
+        )
+
+        # Fill NaNs and fix negative values
+        combo[cat + "_EFF_DIFF"].fillna(0.0, inplace=True)
+        combo.loc[combo[cat + "_EFF_DIFF"] < 0.0, cat + "_EFF_DIFF"] = 0.0
+
+        # Keep first or latest records for further processing
+        if keep == "first":
+            df1 = pd.merge(
+                df1,
+                combo[[cat + "_EFF_DIFF", identifier]],
+                on=identifier,
+            )
+        else:
+            df2 = pd.merge(
+                df2,
+                combo[[cat + "_EFF_DIFF", identifier]],
+                on=identifier,
+            )
+
+    if keep == "first":
+        df = df1
+    else:
+        df = df2
```

As I said in the `upgrades.py` file, I think this logic makes room for error:
1. Imagine that in one of these `if` statements, you misspell "first" (e.g. "fist"). It would be a silent error and `df2` would always be used.
2.`df1` and `df2` are very general;